gen/builders/tipsets.go

```diff
@@ -0,0 +1,79 @@
+package builders
+
+import (
+	"github.com/filecoin-project/test-vectors/schema"
+
+	"github.com/filecoin-project/specs-actors/actors/abi"
+
+	"github.com/filecoin-project/go-address"
+)
+
+// TipsetSeq is a sequence of tipsets to be applied during the test.
+// TipsetSeq provides methods to build a sequence of tipsets, such as
+// adding new tipsets and null rounds.
+type TipsetSeq struct {
+	tipsets  []*Tipset
+	epoch    abi.ChainEpoch
+	messages []*ApplicableMessage
+}
+
+type Tipset struct {
+	tss *TipsetSeq
+	schema.Tipset
+}
+
+type Block = schema.Block
+
+// NewTipsetSeq returns a new TipSetSeq object initialized at the provided
+// epoch.
+func NewTipsetSeq(initialEpoch abi.ChainEpoch) *TipsetSeq {
+	return &TipsetSeq{epoch: initialEpoch}
+}
+
+// All returns all tipsets that have been registered.
+func (tss *TipsetSeq) All() []*Tipset {
+	return tss.tipsets
+}
+
+// Messages returns all ApplicableMessages that have been registered, in the
+// same order they were registered.
+func (tss *TipsetSeq) Messages() []*ApplicableMessage {
+	return tss.messages
+}
+
+// Next enrols a new Tipset, with the supplied base fee, and advances the epoch
+// by 1.
+func (tss *TipsetSeq) Next(baseFee abi.TokenAmount) *Tipset {
+	ts := &Tipset{
+		tss: tss,
+		Tipset: schema.Tipset{
+			Epoch:   tss.epoch,
+			BaseFee: baseFee,
+		},
+	}
+	tss.tipsets = append(tss.tipsets, ts)
+	tss.epoch++ // advance the epoch.
+	return ts
+}
+
+// NullRounds enrols as many null rounds as indicated, advancing the epoch by
+// the count.
+func (tss *TipsetSeq) NullRounds(count uint64) {
+	tss.epoch += abi.ChainEpoch(count)
+}
+
+// Block adds a new block to this tipset, produced by the indicated miner, with
+// the supplied wincount, and containing the listed (and previously staged)
```

Comment:
this comment claims messages are already staged in the TipsetSeq before Block is called, but the method then appends them.
Something is smelly here.
Maybe `TipsetSeq` should have a method that should both add a block to the last tipset, and learn about the messages in that block?

Reply:
By staged I mean that the message is an `ApplicableMessage`, which can only be generated through a `Messages` object. The Tipset do their own management of messages. For example, you can stage 100 messages, but only actually incorporate 10 to blocks, that's totally fine.